### PR TITLE
User/rluu/path pattern search

### PR DESCRIPTION
**Pull request by ryanluu12345**

### CODE REVIEW

File: metrics/http.go [hunk 1]
```diff
@@ -20,9 +23,58 @@ type (
...
+}
+
+// InitMetricDetails includes details about the server host, status codes, and path/verb for each endpoint.
+// This is used to figure out which specific metric combination to initialize.
+type InitMetricDetails struct {
+	// EndpointDetails are the HTTP details for each Path + Verb combo.
+	EndpointDetails []*HTTPEndpointDetails
+	// Host is the host of the actual running server.
+	Host string
+	// StatusCodes are the set of status codes that are possible for the endpoints (i.e. 400, 500, 200)
```
Comment (raphael) [inline metrics/http.go:41]: That breakdown does not seem quite right: different endpoints may return different status codes.

File: metrics/http.go [hunk 2]
```diff
@@ -73,6 +132,11 @@ func HTTP(ctx context.Context) func(http.Handler) http.Handler {
 			h.ServeHTTP(rw, req)
 
 			labels[labelHTTPStatusCode] = strconv.Itoa(rw.StatusCode)
+			val := ctx.Value(goa.PathPatternKey)
+			fmt.Println(val)
```
Comment (raphael) [inline metrics/http.go:136]: Looks like debug cruft?

**Review by raphael — COMMENTED** [review]
Hello, thank you for the PR!. Can you explain the problem this is intended to solve? There are a number of changes that would have to be made before this can be merged (left a few comments) but more importantly we should have a little discussion first around the intent and how best to approach it...

**Comment by ryanluu12345** [comment]
My apologies here! I thought I had opened a PR to my fork. I was just experimenting with different approaches to use the path pattern (`/api/v1/connections/{id}`) instead of the full path (`/api/v1/connections/1234`) for metrics to avoid explosion because of cardinality.

I ended up resolving this by making a lookup of paths to patterns from the `Server.MountPoints` and then using a `metrics.WithRouteResolver`. This did not require any changes to `clue` itself.

Just for future reference, I'm wondering if in the root `goadesign/goa` repository, there are plans to make the path pattern accessible in context so that middlewares have access to it? I could see this being useful for metrics, tracing, etc.

File: metrics/http.go (hunk 1 above)
Comment (ryanluu12345) [reply metrics/http.go:41]: Yes, you're correct about this. I think what I would have done to change this is make each endpoint detail have its own relevant status codes. This was originally meant to solve the problem of metrics not being initialized for each label combination (status + HTTP path)

File: go.mod [hunk 3]
```diff
@@ -58,3 +58,5 @@ require (
 	google.golang.org/genproto/googleapis/rpc v0.0.0-20230629202037-9506855d4529 // indirect
 	gopkg.in/yaml.v3 v3.0.1 // indirect
 )
+
+replace goa.design/goa/v3 => github.com/ryanluu12345/goa/v3 v3.0.0-20230806062212-3626f267e8cd
```
[the remark this answers is not in the document]
Comment (ryanluu12345) [reply go.mod:62]: Yeah good point, it was just experimental code I shouldn't have made a PR for

**Comment by raphael** [comment]
Ah that makes sense! The initial intent was to keep the initialization code for metrics outside of Clue's purview given how use-case specific that can be but always open to suggestions.

Wrt getting the design path into the request context, I could see this being something the generated Goa handlers add, however that would mean that the values would only be available after the service method has run. Another approach that's doable today is to wrap the Goa muxer with a custom one whose `Handle` method would inject the path pattern in the request context. Something like:

```Go
func InstrumentMux(mux goahttp.Muxer) goahttp.Muxer {
    return &instrumentedMux{mux}
}

type instrumentedMux  struct {
    mux goahttp.Muxer
}

func Handle(verb, pattern string, h http.HandlerFunc) {
    f = func(w http.ResponseWriter, r *http.Request) {
        r = r.WithContext(r.Context().WithValue("pattern", pattern))
        h(w, r)
    }
}
```

(note: I did not try the code above, probably doesn't work as is :) )

**Comment by ryanluu12345** [comment]
> 

Hey @raphael , thank you so much for the thorough reply, explaining both your thought process for the metrics initializing and the way we can make an instrumented mux that can get the path. I just wanted to address both of these points (apologies for the longer text below):

1. Instrument mux somewhat works, but not completely with `clue/metrics`
I ended up doing a diff that looks similar to what you showed above, although more fleshed out to adhere to the interface for the goahttp muxer. The resulting code is as follows:
```
type instrumentedMux struct {
	mux goahttp.Muxer
}

func InstrumentMux(mux goahttp.Muxer) goahttp.Muxer {
	return &instrumentedMux{
		mux: mux,
	}
}

func (im *instrumentedMux) Vars(r *http.Request) map[string]string {
	return im.mux.Vars(r)
}

func (im *instrumentedMux) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	im.mux.ServeHTTP(w, r)
}

func (im *instrumentedMux) Handle(verb, pattern string, h http.HandlerFunc) {
	f := func(w http.ResponseWriter, r *http.Request) {
		ctx := context.WithValue(r.Context(), "pattern", pattern)
		r = r.WithContext(ctx)
		h(w, r)
	}
	im.mux.Handle(verb, pattern, f)
}
```
What I notice is that when I attempt to retrieve the pattern from one of the service handler methods (`goahttp.endpoint` adhering handlers), it can correctly pick up the path pattern: `ctx.Value("pattern")` returns the valid path. I also noticed that in `HTTP middleware` that I write myself and register to `server.Server`, I can also pull the value out of context. 

**However**, it appears that when I try to get the value out of the `clue` metrics handler, it shows up as nil every time even though I'm using the same pattern extraction pattern: `ctx.Value("pattern")`. My assumption is that for some reason the request context is getting lost between all the middleware/handlers for the service running and the one that clue runs to report metrics.

For context, here is how I'm registering the clue metrics handler. Does it have something to do with wrapping around the instrumented mux like how it's done? Or is it a new context entirely that overrides it? I'm a bit confused here.
```
// Wraps the handler in the middleware for Prometheus metrics.
handler := metrics.HTTP(ctx)(iMux)

registerMiddlewares(ctx, s.logger, svr)
server.Mount(iMux, svr)

return &http.Server{
  Addr:    fmt.Sprintf("0.0.0.0:%d", cfg.ListenPort),
  Handler: handler,
}, nil
```

2. Regarding metrics initializing, I think it makes sense to leave dynamic init to the user themselves. However, I see that in order to get access to the metrics, one needs to unpack it from context with a key that is private by default. This makes it challenging to access the metrics from state and initialize them on our end. I can see the case for keeping this private because it means no one can tamper with the metrics, but it also means people can't use the metrics and init as they see fit. Maybe I'm misunderstanding what's available, but that's my understanding after a quick read of `metrics/context.go`.

If metrics init can be plumbed in, then probably packing some details in context about exact routes/status code/verb combinations and specifying which metric to init for would be a good option. It could be similar to the spirit of some of this PR without the breaking change.

**Comment by raphael** [comment]
Sorry for the late reply. The problem with the context missing the pattern is probably due to the fact that the metrics handler runs first (outer most) then the instrumented mux runs and adds the pattern to the context. Instead the code could call the mux `Use` passing in the metrics middleware - this would add the metrics handler downstream of the instrumented mux so it should have access to any data that the mux added to the context.

**Comment by raphael** [comment]
Closing this PR as more discussion is probably needed.

**Pull request closed without merging**